// topos/lib/test-connections.js
#!/usr/bin/env node

'use strict';

var connectionTester = require('connection-tester'),
    Table = require('easy-table'),
    colors = require('colors'),
    url = require('url'),
    async = require('async');

// Default connectionCheckTimeout
var CONNECT_TIMEOUT = 1500;

function extractHostPort(appTopo, topoProp) {
    var endPoint =  appTopo[topoProp],
        connUrl,
        ipport,
        ipportHP,
        defaultPort,
        contest = {
            service: topoProp
        };

    if (typeof endPoint === 'object') {
        ipport = endPoint.ipport;
        connUrl = endPoint.url;
        if (ipport) {
            if (ipport.indexOf('^') !== -1) {
                ipport = ipport.split('^')[0];
            }
            ipportHP = ipport.split(':');
            contest.host = ipportHP[1];
            contest.port = ipportHP[2];
        }
        else if (connUrl) {
            connUrl = url.parse(connUrl);
            defaultPort = (connUrl.protocol.indexOf('https') !== -1) ? 443 : 80;

            contest.host = connUrl.hostname;
            contest.port = +connUrl.port || defaultPort;
        }
        else {
            contest.host = endPoint.hostname || endPoint.host || endPoint.ip;
            contest.port = endPoint.port;
        }

        contest.package = endPoint.package || '';
    }

    return contest;
}

function createTestConnectionHandler(contest, timeout) {
    return function(callback) {
        connectionTester.test(contest.host, contest.port, timeout, function(err, output) {
            contest.success = output.success === true ? '√'.green : 'X'.red;
            contest.error = output.error ? colors.red(output.error) : '';
            callback(null, contest);
        });
    };
}

function extractConnectionInfos(data){

    var contests = [];

    //Level 1 of data in App
    for (var appProp in data) {
        var appTopo = data[appProp];
        if (typeof appTopo === 'object') {
            if (appTopo.host || appTopo.port || appTopo.ipport || appTopo.url) {
                contests.push(extractHostPort(data, appProp));
            } else {
                for (var topoProp in appTopo) {
                    contests.push(extractHostPort(appTopo, topoProp));
                }
            }
        }
    }

    return contests;
}

exports = module.exports = {
    test: function TestConnections(data, connectTimeout) {

        var testHandlers = [];

        var contests = extractConnectionInfos(data);

        var results = [];

        var timeout = isNaN(connectTimeout) ? CONNECT_TIMEOUT : connectTimeout;

        // synchronous calls
        contests.forEach(function(contest) {
            if (contest.host || contest.port) {
                var testOutput = connectionTester.test(contest.host, contest.port, timeout);
                contest.success = testOutput.success === true ? '√'.green : 'X'.red;
                contest.error = testOutput.error ? colors.red(testOutput.error) : '';
                results.push(contest);
            }
        });

        return results;
    },

    testAsync: function TestConnections(data, connectTimeout, callback) {

        var testHandlers = [];

        var contests = extractConnectionInfos(data);

        var timeout = isNaN(connectTimeout) ? CONNECT_TIMEOUT : connectTimeout;

        // Create connection test handlers
        contests.forEach(function(contest) {
            if (contest.host || contest.port) {
                testHandlers.push(createTestConnectionHandler(contest, timeout));
            }
        });

        async.parallel(testHandlers, function(err, results) {
            if (callback) {
                return callback(err, results);
            }
        });
    },

    print: function Print(results) {
        var t = new Table();

        results.forEach(function (contest) {
            t.cell('     ', '     ');
            t.cell('service', contest.service);
            t.cell('package', contest.package);
            t.cell('host', contest.host);
            t.cell('port', contest.port);
            t.cell('status', contest.success);
            t.cell('reason', contest.error);
            t.newRow();
        });

        console.log();
        console.log('\t\t', 'TOPO Connection Validator'.grey.underline);
        console.log(t.toString());
        console.log();
    }
};
